### PR TITLE
Add extra information when failing to parse contract

### PR DESCRIPTION
**Pull request by velo**
Inspired by #1173

I decided to add some extra information to contract parsing... to be attached to exceptions.

So, instead of just saying `Method has too many Body parameters` go figure what that means

I attached some extra bits:

```
java.lang.IllegalStateException: Method has too many Body parameters: public abstract feign.Response feign.DefaultContractTest$MixedAnnotations.findAllClientsByUid2(java.lang.String,java.lang.Integer,java.lang.Integer)
Warnings:
- Class MixedAnnotations has no annotations, it may affect contract Default
- Parameter String has annotations [Category] that are not used by contract Default
- Parameter Integer has no annotations, it may affect contract Default
	at feign.Util.checkState(Util.java:129)
	at feign.Contract$BaseContract.parseAndValidateMetadata(Contract.java:124)
	at feign.Contract$BaseContract.parseAndValidateMetadata(Contract.java:62)
	at feign.DeclarativeContract.parseAndValidateMetadata(DeclarativeContract.java:38)
...
```

probably needs more work =)
and introduces more mutability to `MethodMetadata` =(((

### CODE REVIEW

File: core/src/main/java/feign/DeclarativeContract.java [hunk 1]
```diff
@@ -45,10 +47,30 @@
...
-        .forEach(annotation -> classAnnotationProcessors.stream()
-            .filter(processor -> processor.test(annotation))
-            .forEach(processor -> processor.process(annotation, data)));
+    if (Arrays.stream(targetType.getAnnotations())
+        .anyMatch(annotation -> classAnnotationProcessors.stream()
+            .anyMatch(processor -> processor.test(annotation)))) {
+      Arrays.stream(targetType.getAnnotations())
+          .forEach(annotation -> classAnnotationProcessors.stream()
+              .filter(processor -> processor.test(annotation))
+              .forEach(processor -> processor.process(annotation, data)));
```
Comment (kdavisk6) [inline core/src/main/java/feign/DeclarativeContract.java]: Keying off my previous comment, the end result here is 

```java
Collection<AnnotationProcessor> applicableProcessors = this.getAnnotationProcessorsFor(targetType);
applicableProcessors.forEach(processor -> process(annotation, data));
```

File: core/src/main/java/feign/DeclarativeContract.java [hunk 2]
```diff
@@ -60,9 +82,19 @@ protected final void processAnnotationOnClass(MethodMetadata data, Class<?> targ
   protected final void processAnnotationOnMethod(MethodMetadata data,
                                                  Annotation annotation,
                                                  Method method) {
-    methodAnnotationProcessors.stream()
-        .filter(processor -> processor.test(annotation))
-        .forEach(processor -> processor.process(annotation, data));
+    if (methodAnnotationProcessors.stream()
+        .anyMatch(processor -> processor.test(annotation))) {
```
Comment (kdavisk6) [inline core/src/main/java/feign/DeclarativeContract.java]: Similar approach here, maybe create a function to do this test.

File: core/src/main/java/feign/DeclarativeContract.java [hunk 3]
```diff
@@ -45,10 +47,30 @@
    */
   @Override
   protected final void processAnnotationOnClass(MethodMetadata data, Class<?> targetType) {
-    Arrays.stream(targetType.getAnnotations())
-        .forEach(annotation -> classAnnotationProcessors.stream()
-            .filter(processor -> processor.test(annotation))
-            .forEach(processor -> processor.process(annotation, data)));
+    if (Arrays.stream(targetType.getAnnotations())
+        .anyMatch(annotation -> classAnnotationProcessors.stream()
+            .anyMatch(processor -> processor.test(annotation)))) {
```
[the remark this answers is not in the document]
Comment (velo) [reply core/src/main/java/feign/DeclarativeContract.java]: I couldn't make this happen in a meaningful way.

There are 2 loops, so can't really get a single list that I can just use w/o further filtering.

[the remark this answers is not in the document]
Comment (kdavisk6) [reply core/src/main/java/feign/DeclarativeContract.java]: Not asking you to simplify the loop, just move it to a function so it is a little more readable.

**Pull request merged**